### ISSUE TITLE
obabel with the --separate option produces incomplete results when input data is read from stdin

### ISSUE THREAD

**Issue by vaitkus** [openbabel/openbabel#2386]
- [x] I believe this to be a bug with Open Babel
- [ ] This is a feature request

## Environment Information
Open Babel version: 3.1.1
Operating system and version: Linuxmint-20.1

## Expected Behavior

The program should read input data from _stdin_ and output all separate disconnected fragments into individual molecular records.

## Actual Behavior

The program reads input data from _stdin_ and outputs only one of the disconnected fragments as an individual molecular record.

## Steps to Reproduce

I have observed this behaviour with SMILES and SDF formats, but it is possible that other formats are affected as well.

Contents of the `test.smi` file:
<pre>
[Na].[Cl]
</pre>

Providing the `test.smi` file directly results in the expected output:
<pre>
$ obabel -iSMI test.smi -oSMI --separate

[Na]	#1
[Cl]	#2
2 molecules converted
</pre>

However, reading the same data from _stdin_ results in only the first component being output:
<pre>
$ cat test.smi | obabel -iSMI -oSMI --separate

[Na]	#1
1 molecule converted
</pre>

**Comment by nbehrnd** [comment]
There are at least two solutions to resolve the issue reported.  Either

- with file `test.smi` you present, provide the instruction in a *different sequence*, i.e.

  `obabel -ismi test.smi -osmi --separate`

  because `-ismi` instructs OpenBabel the (then in the *following* defined) input file consists of content of the .smi syntax.  This instruction actually works regardless if the file in question actually possesses the file extension `.smi`, or not.

.or.

- if you like to stay on the command line and don't want to create a `.smi` file in advance, type

  `obabel -:'[Na].[Cl]' -osmi --separate`

  to provide the SMILES string directly.  *Caveat:* The instruction will fail if there is a space between the colon and the leading quote sign.  If you want, you may enclose the SMILES string with double quotes

  `obabel -:"[Na].[Cl]" -osmi --separate`

  to generate the same output.  And yes, `-osmi` and `-oSMI` seem to be functionally equivalent to each other, too.

As checked (OpenBabel 3.1.1, provided by the repositories of Linux Debian 11 / bullseye, branch testing), both approaches yield the result you anticipate:
```
[Na]	#1
[Cl]	#2
2 molecules converted 
```

**Comment by e-kwsm** [comment]
I’m not sure why but obabel converts two molecules if two or more blank lines are added:

```shellsession
$ obabel -V
Open Babel 3.1.1 -- Jun 14 2021 -- 01:51:20
$ for n in {0..2}; do perl -E "say '[Na].[Cl]' . \"\n\"x$n" | obabel -ismi -osmi --separate; done
[Na]    #1
1 molecule converted
[Na]    #1
1 molecule converted
[Na]    #1
[Cl]    #2
2 molecules converted
```

**Comment by nbehrnd** [comment]
@vaitkus  True, the first suggestion boils down to not use `cat` and `|`.  When I moderate OpenBabel with Python's `subprocess` module, using `Popen` to work with input data which are syntactically wrong, I may fetch `stderr` (see example `check.py` in the example archive attached below).  The translation of this approach to cover your instruction does not yield an observable error message (see `observation_vaitkus.py`).  I don't know if an other language (of your choice) is going to be more verbose / more suitable here.

The suggest "different sequence" referred to replace the approach with cat (get the content of the file, pipe this into subsequently started OpenBabel) by starting OpenBabel in first place (and let OpenBabel fetch the content of `test.smi`).  This is the only change in the sequence of events I put on the table.

[convert_io.zip](https://github.com/openbabel/openbabel/files/6669604/convert_io.zip)

**Comment by nbehrnd** [comment]
@e-kwsm Affirmative, it is an interesting observation.  Addition of two blank lines in the input file alters OpenBabel's output from
```
[Na]	#1
1 molecule converted
```
into
```
[Na]	#1
[Cl]	#2
2 molecules converted
```
[test_b.smi.txt](https://github.com/openbabel/openbabel/files/6669683/test_b.smi.txt)

**Comment by ashaymardanov** [comment]
The same behavior can be confirmed for any popular molecular format. Just do something like this
`diff -y <(obabel -:"[Na].[Cl]" -omol2 --separate) <(obabel -:"[Na].[Cl]" -omol2 | obabel -imol2 -omol2 --separate)`

or this

```
for i in smi mol2 hin sdf pdb; do 
	echo "----------${i}---------"; 
	obabel -:"[Na].[Cl]" -O "/tmp/mol.${i}"; 
	obabel "/tmp/mol.${i}" -O "/tmp/mol_sep.${i}" --separate; 
	cat "/tmp/mol.${i}" | obabel -i ${i} -o ${i} --separate | diff -y "/tmp/mol_sep.${i}" - ; 
	echo; echo; 
done 2>/dev/null
```
and see the difference.

I've checked this for `Open Babel 2.3.2 -- Jul 26 2017 -- 20:11:55` and `Open Babel 3.1.1 -- Jun 10 2021 -- 04:27:50` and both act the same.

**Comment by vaitkus** [comment]
Can still reproduce this with obabel built from the latest commit (https://github.com/openbabel/openbabel/commit/889c350feb179b43aa43985799910149d4eaa2bc).
